roll/ast.py

```diff
@@ -0,0 +1,685 @@
+
+
+class TokenFunction(Token):
+    def __init__(self, arg_name, expression, arg_id=None):
+        self.arg_name = arg_name
+        self.arg_id = arg_id if arg_id is not None else hash(arg_name)
+        self.expression = expression
+        logging.debug(self.__class__.__name__, self)
+
+    @trace
+    def reduce(self, env, counter):
+        return self
+
+    def substitute(self, old_to_new):
+        # If the function arg hides a variable in the map, remove it
+        if self.arg_id in old_to_new:
+            old_to_new = old_to_new.copy()
+            del old_to_new[self.arg_id]
+        # Copy the function body
+        new_expr = self.expression.substitute(old_to_new)
+        # Return reconstructed function statement
+        return TokenFunction(self.arg_name, new_expr, self.arg_id)
+
+    def type(self, env):
+        pass
+
+    @trace
+    def hash_vars(self, counter, map):
+        new_map = map.copy()
+        new_map[self.arg_name] = counter.next_id
+        self.arg_id = counter.next_id
+        counter.pop_id()
+        self.expression.hash_vars(counter, new_map)
+
+    def __str__(self):
+        return f"({self.arg_name}->{self.expression})"
+
+
+class TokenApplication(Token):
+    def __init__(self, lhs, rhs):
+        self.lhs = lhs
+        self.rhs = rhs
+        logging.debug(self.__class__.__name__, self)
+
+    @trace
+    def reduce(self, env, counter):
+        # Apply as many arguments as possible (assumes the application is valid)
+        # May result in a partially-applied function
+        # Create declarations from lhs function(s)
+        out = self.lhs.dereference(env)
+        decls = []
+        for expr in self.rhs:
+            decls.append(Assignment(out.arg_name, expr, id=out.arg_id))
+            out = (
+                out.expression
+            )  # Note: this used to attempt to .dereference() - I can't remember why but it seems to work without it
+        # Substitute variables for scoped variables (allows recursion)
+        substitutions = {}
+        env = env.copy()
+        for decl in decls:
+            substitutions[decl.identifier] = counter.next_scope_id
+            env.closure[counter.next_scope_id] = decl.expression
+            counter.pop_scope_id()
+        out = out.substitute(substitutions)
+        # Check if the application was partial
+        # If it was, extract the rest of the function arguments and put the Let token at the bottom
+        if isfunction(out):
+            func_path = [out]
+            out = out.expression.dereference(env)
+            while isfunction(out):
+                func_path.append(out)
+                out = out.expression.dereference(env)
+            func_path[-1] = TokenFunction(
+                func_path[-1].arg_name,
+                TokenLet(decls, out),
+                func_path[-1].arg_id - 1000,
+            )
+            for i in range(len(func_path) - 1, 0, -1):

```

Comment:
This might be nicer by using https://docs.python.org/3/library/functions.html#reversed and https://docs.python.org/3/library/functions.html#enumerate